theano/gpuarray/c_code/ctc_wrapper.c

```diff
@@ -0,0 +1,296 @@
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "GpuConnectionistTemporalClassification: %s CTC error: %s",
+                      msg,
+                      ctc_msg );
+        return 1;
+    }
+    return 0;
+}
+
+void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr,
+    int ** input_lengths )
+{
+    npy_int num_elements = PyArray_DIMS( input_lengths_arr )[0];
+
+    *input_lengths = (int *) malloc( num_elements * sizeof(int) );
+
+    if ( NULL == (*input_lengths) )
+        return;
+
+    for( npy_int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
+    {
+        (*input_lengths)[elem_idx] = *( (npy_int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
+    }
+}
+
+void create_flat_labels( PyArrayObject * label_matrix, int ** flat_labels,
+    int ** label_lengths )
+{
+    npy_int rows = PyArray_DIMS( label_matrix )[0];
+    npy_int cols = PyArray_DIMS( label_matrix )[1];
+
+    *flat_labels = (int *) calloc( rows * cols, sizeof(int) );
+    if ( NULL == (*flat_labels) )
+        return;
+
+    *label_lengths = (int *) calloc( rows, sizeof(int) );
+    if ( NULL == (*label_lengths) )
+    {
+        free( *flat_labels );
+        *flat_labels = NULL;
+        return;
+    }
+
+    npy_int label_index = 0;
+    for( npy_int row_idx = 0; row_idx < rows; ++row_idx )
+    {
+        npy_int label_length = 0;
+        for( npy_int col_idx = 0; col_idx < cols; ++col_idx )
+        {
+            npy_int label = *( (npy_int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
+            if ( label >= 0 )  // negative values are assumed to be padding
+            {
+                (*flat_labels)[ label_index++ ] = label;
+                ++label_length;
+            }
+        }
+        (*label_lengths)[ row_idx ] = label_length;
+    }
+}
+
+#section support_code_apply
+
+int APPLY_SPECIFIC(ctc_cost_gpu)(PyGpuArrayObject   *  in_activations,
+                                 PyArrayObject      *  in_labels,
+                                 PyArrayObject      *  in_input_lengths,
+                                 PyGpuArrayObject   ** out_costs,
+                                 PyGpuArrayObject   ** out_gradients,
+                                 PyGpuContextObject *  gpu_context)
+{
+    ctc_context_t ctc_object;
+    ctc_context_t * context = &ctc_object;
+
+    size_t gpu_workspace_size;
+    int ctc_error = 0;
+
+    const size_t num_activations = PyGpuArray_DIMS( in_activations )[0];
+    const size_t minibatch_size = PyGpuArray_DIMS( in_activations )[1];
+    const size_t alphabet_size = PyGpuArray_DIMS( in_activations )[2];
+    const size_t cost_size = minibatch_size;
+
+    const size_t grad_dims[3] = { num_activations, minibatch_size, alphabet_size };
+
+    float * costs = NULL,
+          * activations = NULL,
+          * gradients = NULL;
+
+    cuda_enter( gpu_context->ctx );
+
+    ctc_context_init( context, gpu_context );
+
+    switch (in_activations->ga.typecode)
```

Comment:
I think it does not apply here, since we're calling ``gpu_contiguous(activations)`` in ``make_node``, prior to calling the C code.